src/compiler/scala/tools/nsc/ast/parser/MatchRewriter.scala

```diff
@@ -0,0 +1,113 @@
+package scala.tools.nsc
+package ast.parser
+
+class MatchRewriter[G <: scala.tools.nsc.Global](val global: G) {
+  import global._
+
+  /** Entry point. */
+  object rewrite extends Transformer {
+    def apply(t: Tree): Tree = {
+      logResult("Transforming: " + t)(transform(t))
+    }
+    override def transform(t: Tree): Tree = t match {
+      case m: Match => new TranslateMatch(super.transform(m).asInstanceOf[Match]) resultTree
+      case _        => super.transform(t)
+    }
+  }
+
+  object CaseDecompose {
+    /** That's right, the compiler generates a wrong AppliedTypeTree
+     *  instead of a TypeApply tree when there are type args in a pattern.
+     *  "We can rebuild it... we have the technology."
+     *
+     *  Unfortunately fixing it at the source revealed many other
+     *  points in the compiler which depend on things happening as
+     *  they presently do. You, dear reader, are welcome to try it.
+     */
+    def makeTermTree(t: Tree): Tree = t match {
+      case AppliedTypeTree(fun, args) => TypeApply(makeTermTree(fun), args)
+      case Select(qual, name)         => Select(makeTermTree(qual), name.toTermName)
+      case Ident(name)                => Ident(name.toTermName)
+    }
+    def unapply(t: Tree) = t match {
+      case CaseDef(Apply(AppliedTypeTree(fun, targs), patternVars), guard, expr) =>
+        Some((makeTermTree(fun), targs, patternVars, guard, expr))
+      case CaseDef(Apply(fn, patternVars), guard, expr) =>
+        Some((fn, Nil, patternVars, guard, expr))
+      case CaseDef(pat, guard, expr) =>
+        Some((pat, Nil, Nil, guard, expr))
+      case _ =>
+        None
+    }
+  }
+
+  class TranslateMatch(m: Match) {
+    val Match(expr, cases) = m
+
+    /** The list of valdefs to prepend to the rewritten match. */
+    var valDefs = Vector[Tree]()
+
+    /** Memoize the scrutinee expression. */
+    lazy val exprVal = currentUnit.freshTermName("Scrutinee")
+    lazy val expr1   = q"val $exprVal = $expr"
```

Comment:
We don't use quasiquotes in the compiler implementation elsewhere, partly because of the lack of ability to the component trees precisely, partly because of the somewhat bloated expansions, and partly keep one fewer dependency in the bootstrap.